highfidelity/hifi
# Don't use blocking connections on the main thread

We're still seeing a lot of deadlocks in the application and crashes in QtScript / QtQml.

Some of this is potentially coming from the `Audio` scripting interface on the main thread calling blocking functions on the `AudioClient` (which runs on it's own thread).  

Some of it is also potentially from threading issues wherever scripts interact with QML in a non-thread safe way, such as in QML based overlays.  

This PR makes several changes.  

* Adds new helper functions and macros for making blocking method invocations.
  * These emit log warning if they occur from the main thread
  * Macro usage ensures that the log warning include the calling function in which the blocking was made
* Migrates _all_ use of `QMetaObject::invokeMethod` that uses `Qt::BlockingQueuedConnection` to use the new methods / macros
* Modified Audio scripting interface to not use blocking functions when communicating with the AudioClient.  
* Modifies Overlay instances to always live on the main thread.  All interaction with the Overlay scripting interface now funnels through the main thread.
* Modifies OvlerayWindows and OverlayWebWindows to live on the main thread.  
* Updates many scripting interfaces to use the self-invocation pattern when they are called from another thread
* Eliminates the unused OverlayPanel code (using the preprocessor so that it can be restored easily if there is some user script taking advantage of it)
* Eliminates support for the tabbed tool window, previously only used by the edit.js code

## Testing

* Verify startup and shutdown function without hanging or crashing
* Verify Overlay functionality still works.  (this [script](https://raw.githubusercontent.com/jherico/DreamingScripts/master/tests/overlays.js) creates a number of overlays)
* Verify audio settings dialog still functions properly (check boxes work, changing input and output devices works)
* Smoke test functionality
* Monitor log output for warnings containing the string `BlockingQueuedConnection` (this doesn't fail the PR, it just informative and should give us clues to additional sources of race conditions)
